### PR TITLE
fix: add app-wide browser security headers for Fastify/SPA

### LINKED ISSUE #1541: Add app-wide browser security headers for Fastify/SPA
Production readiness scan finding.

Report: https://report.first-tree.ai/agent-team-foundation-first-tree-20260709-c47fd3f0.html
Target commit: `65c8feb92be39c3bbd6869febc6e375d15ed35b7`
Severity: High (`needs-check`)

## Problem

The Fastify server serves both API routes and the SPA, but the repo does not show an effective app-wide browser security header layer.

Evidence:

- `packages/server/src/app.ts:607` serves web static files.
- `packages/server/src/app.ts:612-629` registers static file serving and SPA fallback.
- Repo search found only attachment-specific `X-Content-Type-Options: nosniff` in `packages/server/src/api/attachments.ts:54`.
- No effective app-wide CSP / HSTS / frame protection / Referrer-Policy / global nosniff setter was found in app source, Dockerfile, or workflows.

An edge/CDN may already set these headers out-of-repo, but the repo does not make that contract visible or testable.

## Suggested fix

Register app-wide secure headers before API/static routes, for example with `@fastify/helmet` or equivalent hooks.

Minimum headers to cover:

- `Content-Security-Policy` or at least `frame-ancestors`
- `Strict-Transport-Security`
- `X-Content-Type-Options: nosniff`
- `Referrer-Policy`
- `X-Frame-Options` if not covered by CSP `frame-ancestors`

If CapRover/Cloudflare owns these headers, commit the edge header policy or add a deployment check so the guarantee is visible.

## Verification

```bash
curl -sI https://<prod-host>/ | grep -iE 'content-security-policy|strict-transport-security|x-content-type-options|referrer-policy|x-frame-options'
```

---

### Description

Static SPA responses are served without a code-defined Content Security Policy, `frame-ancestors`/`X-Frame-Options`, referrer policy, or permissions policy. Inline scripts and remote analytics execute with full origin authority, and the authenticated dashboard can be framed unless external infrastructure happens to compensate.

- **Severity:** Medium
- **Locations:** `packages/web/index.html:26-65`, `packages/server/src/app.ts:607-630`
- **Estimated remediation:** 2-3 days

### Recommended direction

Remove or nonce inline scripts, define and test a least-privilege CSP, disallow framing, and add explicit referrer/permissions headers at the application or verified ingress layer.

### 描述

静态 SPA 响应在提供时没有由代码定义的内容安全策略，也没有 `frame-ancestors`/`X-Frame-Options`、Referrer Policy 或 Permissions Policy。内联脚本和远程分析代码以完整的同源权限执行；除非外部基础设施恰好提供补偿性保护，否则已经过身份验证的仪表盘可以被嵌入框架中。

- **严重性：** 中
- **文件位置：** `packages/web/index.html:26-65`、`packages/server/src/app.ts:607-630`
- **预计修复时间：** 2-3 天

### 建议修复方向

移除内联脚本或为其添加 nonce，定义并测试最小权限 CSP，禁止框架嵌入，并在应用层或经过验证的入口层显式添加 Referrer Policy 和 Permissions Policy 标头。

**Audit date:** 2026-07-12  
**Source revision:** `ea2a9241fd8e128e89236cda2542f180731e3207`

### PR DESCRIPTION
**Pull request by liuchao-001**
Fixes #1541 (and the SEC-041 audit addendum on it).

## Problem

The Fastify server serves both the API and the SPA, and nothing in the repo — app code, Dockerfile, or any edge config — sets browser security headers. No CSP, no frame protection, no HSTS, no global `nosniff` (only the attachments route set one locally), while `index.html` ran three inline scripts (GA4, Clarity, theme init) with full origin authority.

## Approach

One `onSend` hook (`packages/server/src/middleware/security-headers.ts`), registered alongside the existing global body-capture hook **before any route or static registration**, so it covers API JSON, static assets, the SPA shell (direct and fallback), 404/error bodies, and health checks:

- **On every response** — `X-Content-Type-Options: nosniff`, `Referrer-Policy: strict-origin-when-cross-origin`, `X-Frame-Options: DENY`, `Permissions-Policy: camera=(), microphone=(), geolocation=()` (clipboard deliberately not restricted — the app uses it), and `Strict-Transport-Security: max-age=31536000` only when the request arrived over https (`FIRST_TREE_TRUST_PROXY=true` in proxied deployments; plain-HTTP dev never sees it; no `includeSubDomains`/`preload` by design — self-hosted operators on shared parent domains would be the blast radius).
- **On `text/html` responses only** — CSP, staged:
  - enforced immediately (all modes except `off`): `frame-ancestors 'none'; object-src 'none'; base-uri 'self'` — the anti-embedding requirement can only ship enforced because the spec ignores `frame-ancestors` in report-only headers;
  - the full least-privilege policy (script/img/connect/style/font/manifest/form-action) ships as **Content-Security-Policy-Report-Only** by default (`FIRST_TREE_SECURITY_CSP_MODE=report-only`), with `FIRST_TREE_SECURITY_CSP_REPORT_URI` for violation collection, flipping to `enforce` after an observation cycle. `off` is the emergency escape hatch.
  - A side benefit: HTML attachments served `inline` now also carry the CSP, constraining the stored-HTML-attachment XSS surface.
- **Inline scripts externalized** to `packages/web/public/init.js` (theme init first to keep the no-FOUC guarantee, GA4/Clarity order and production hostname gates preserved verbatim), so `script-src` needs no `'unsafe-inline'`. A web-side guard test keeps `index.html` inline-free.

Config lives in the shared server config (`security` block, four `FIRST_TREE_SECURITY_*` envs, documented in `docs/cli-reference.md`); boot logs one line with the active `cspMode`/`hstsEnabled`.

`@fastify/helmet` was evaluated and not used: the two load-bearing behaviors here are conditional (CSP by content type, HSTS by request protocol), which helmet does not express natively, and its COOP default would break the GitHub-App install popup flow — a fully hand-rolled ~small hook is simpler to test than helmet-plus-overrides.

## Implementation note reviewers should know

The hook is written in Fastify's callback (`done`) style, not async: adding a second **async** `onSend` hook exposed a latent race with a fire-and-forget `reply.status(204).send()` in an existing route handler (unhandled `ERR_HTTP_HEADERS_SENT`), reproduced minimally and sidestepped by not adding a microtask window. The pre-existing pattern deserves its own sweep — follow-up issue suggested rather than folded into this PR.

## Verification

- `pnpm check` / `pnpm typecheck` green.
- New `security-headers.test.ts`: **18/18** — exact-value assertions for every unconditional header, CSP scope (HTML yes; `/assets`, API 404, `/healthz` no), report-only vs enforce vs off, connect-src extension and report-uri normalization (directive-injection resistant), HSTS positive/negative including the trust-boundary case (`trustProxy: false` + forged `X-Forwarded-Proto` sends nothing), content-type edge cases.
- Full server package: **239 files / 2581 tests green**. Web package: **193 files / 1569 tests green** (includes the new index.html guard test). `pnpm --filter @first-tree/web build` verified: dist HTML has no inline script; `init.js` passes through byte-identical.
- Rollout runbook (in-repo docs + this PR): keep `report-only` for one release cycle with reports pointed at a Sentry security endpoint, staging first, then flip the default to `enforce`.

---

Designed and reviewed with fire-fable-reviewer (design round 1 absorbed: external avatar evidence for `img-src https:`, HTML-attachment surface, cache semantics, report collection as a hard requirement). Goblet of Fire competition submission — kept as **draft** per competition rules.

### CODE REVIEW

File: packages/web/tests/index-html-no-inline-scripts.test.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * dead in production once the CSP is enforced. The former inline bootstrap
+ * (theme / GA4 / Clarity) lives in `public/init.js`; this test fails the
+ * build the moment someone adds a new inline script or drops the external
+ * bootstrap, instead of the regression surfacing as a broken page.
+ */
+const indexHtml = readFileSync(fileURLToPath(new URL("../index.html", import.meta.url)), "utf8");
+const initJs = readFileSync(fileURLToPath(new URL("../public/init.js", import.meta.url)), "utf8");
+// Browsers ignore markup inside comments, so the guard should too — the
+// prose comment above the bootstrap tag legitimately mentions "<script>".
+const markup = indexHtml.replace(/<!--[\s\S]*?-->/g, "");
```
[the remark this answers is not in the document]
Comment (liuchao-001) [reply packages/web/tests/index-html-no-inline-scripts.test.ts]: Addressed in 984049026: the guard no longer strips comments with a single-pass regex. It now parses index.html with DOMParser (happy-dom) and asserts on real `<script>` elements, so commented-out markup is handled by the HTML parser itself and the flagged pattern is gone entirely.

For context: the flagged code was a test-only scan of the repo's own static index.html — never untrusted input, and the result is never rendered — so there was no exploitable path. The parser-based guard is simply stronger (browser-faithful comment handling) and CodeQL-clean. Mutation-checked: an injected inline `<script>` still fails the guard. Alert #96 should auto-close on the next analysis of this branch.
